src/arthur_common/aggregations/functions/agentic_aggregations.py

```diff
@@ -0,0 +1,727 @@
+from typing import Annotated
+from uuid import UUID
+import json
+from datetime import datetime
+
+import pandas as pd
+from duckdb import DuckDBPyConnection
+
+from arthur_common.aggregations.aggregator import (
+    NumericAggregationFunction,
+    SketchAggregationFunction,
+)
+from arthur_common.models.datasets import ModelProblemType
+from arthur_common.models.metrics import DatasetReference, NumericMetric, SketchMetric
+from arthur_common.models.schema_definitions import (
+    DType,
+    MetricColumnParameterAnnotation,
+    MetricDatasetParameterAnnotation,
+    ScalarType,
+    ScopeSchemaTag,
+)
+
+# Global threshold for pass/fail determination
+RELEVANCE_SCORE_THRESHOLD = 0.5
+TOOL_SCORE_PASS_VALUE = 1
+TOOL_SCORE_NO_TOOL_VALUE = 2
+
+
+def extract_spans_with_metrics_and_agents(root_spans):
+    """Recursively extract all spans with metrics and their associated agent names from the span tree.
+    
+    Returns:
+        List of tuples: (span, agent_name)
+    """
+    spans_with_metrics_and_agents = []
+    
+    def traverse_spans(spans, current_agent_name='unknown'):
+        for span in spans:
+            # Update current agent name if this span is an AGENT
+            if span.get('span_kind') == 'AGENT':
+                try:
+                    raw_data = span.get('raw_data', {})
+                    if isinstance(raw_data, str):
+                        raw_data = json.loads(raw_data)
+                    
+                    # Try to get agent name from the span's name field
+                    agent_name = raw_data.get('name', 'unknown')
+                    if agent_name != 'unknown':
+                        current_agent_name = agent_name
+                except (json.JSONDecodeError, KeyError, TypeError):
+                    pass
+            
+            # Check if this span has metrics
+            if span.get('metric_results') and len(span.get('metric_results', [])) > 0:
+                spans_with_metrics_and_agents.append((span, current_agent_name))
+            
+            # Recursively traverse children with the current agent name
+            if span.get('children'):
+                traverse_spans(span['children'], current_agent_name)
```

Comment:
are we guaranteed that the `span.children` attribute will always be an empty list? Otherwise worth wrapping this in a `span.get('children', [])`